### DISCUSSION TITLE
Showing the Output Channel after Debug Console does not refocus

### DISCUSSION THREAD

**Discussion by vvvjo** [eclipse-theia/theia#8085] [General]
Hello, 

I'm working on an extension that should be compatible with both VSCode and Theia. I noticed a difference in behavior when using `debug.startDebugging` and then `OutputChannel.instance.show` functions. In VSCode, the `OutputChannel.instance.show` will override the panel that is revealed but in Theia the focus to the OutputChannel will not refocus. Is this working as intended or a bug in Theia? If it is working as intended, how would we refocus back to the Output panel programmatically?  

Below is the sample code that I used to repro and the corresponding GIF.

Thank you. 

Code used:
```typescript
import * as theia from '@theia/plugin';

export function start(context: theia.PluginContext) {
    context.subscriptions.push(theia.commands.registerCommand({
        id: 'start-debugging',
        label: 'Hello World: Start Debugging'
    }, async (...args: any[]) => {
        const testChannel = theia.window.createOutputChannel('test-channel');
        testChannel.appendLine('starting...');
        await theia.debug.startDebugging(theia.workspace.workspaceFolders![0], {
            "type": "node",
            "request": "launch",
            "name": "Launch Program",
            "program": "${workspaceFolder}/theia-hello-world-plugin/node-test/index.js",
        });
        testChannel.appendLine('running...');
        testChannel.appendLine('running...');
        testChannel.appendLine('running...');
        testChannel.show(); 
    }));
```

Theia:
![issue](https://user-images.githubusercontent.com/13400593/85603674-70b19780-b605-11ea-9de2-b59a1ccaf39b.gif)

**Comment by kittaakos** [comment]
What's your Theia version, @vincentjocodes?

**Reply by kittaakos** [reply]
Hm, I think the problem is that we do not support the `preserveFocus` for `OutputChannel#show`. It should be `show(preserveFocus?: boolean): void;`.

**Reply by kittaakos** [reply]
It seems to be there, please try with `show(true)`.

**Reply by vvvjo** [reply]
Hello @kittaakos, 

Thank for your reply. My Theia version is at `@theia/example-browser 1.2.0`. 

I have just tried your suggestion setting `OutputChannel#show(true)` but I get the same result. Anything else you could suggest? 

Only way I was able to show the Output panel correctly was by adding `debug.internalConsoleOptions: "neverOpen"` to the launch configuration. However, this does not show the Debug Console at all. Is this the preferred method in Theia or is there an issue here that needs to be fixed?

Thanks!

**Reply by vvvjo** [reply]
@kittaakos I want to follow up on this. Thanks!

**Reply by kittaakos** [reply]
Can you please set up a repo with an example?

**Reply by kittaakos** [reply]
> My Theia version is at `@theia/example-browser 1.2.0`.

Can you please try with `1.3.0` before you prepare any repo? As I see some related [changes](https://github.com/eclipse-theia/theia/commit/60437227287773d7b4e032ce4b6e2568cbb87d65) were added after the `1.2.0` release.

**Reply by vvvjo** [reply]
@kittaakos I tried with `1.3.0` but I got the same results. Here is the example Theia extension: https://github.com/vincentjocodes/theia-hello-world-plugin. Thanks!

**Reply by kittaakos** [reply]
I could reproduce it, thanks for the example. I will look into it later today. I think the problem is the following: you will call a synchronous `show`, which will end up calling this code:

https://github.com/eclipse-theia/theia/blob/68a3453ed1de6279dc8a44e7d846b92a0953ebe9/packages/plugin-ext/src/plugin/output-channel/output-channel-item.ts#L49-L52

`show` will call an asynchronous `reveal` here:

https://github.com/eclipse-theia/theia/blob/68a3453ed1de6279dc8a44e7d846b92a0953ebe9/packages/plugin-ext/src/main/browser/output-channel-registry-main.ts#L63-L71

**Reply by kittaakos** [reply]
First off, I am not that familiar with the Theia plugin API, so I converted your code into a real VS Code extension and tried that. Here is the repo, if you're interested: https://github.com/kittaakos/output-channel-example

I found that for some reason, the very first debug session behaves differently on the UI:
https://github.com/eclipse-theia/theia/blob/68a3453ed1de6279dc8a44e7d846b92a0953ebe9/packages/debug/src/browser/debug-frontend-application-contribution.ts#L450

Maybe @akosyakov knows why and can advise what to do to align the behavior with VS Code.

Apparently, VS Code does not distinguish between debug sessions. After changing the initial `reveal` to `false`, I had the same UX as in code. Feel free to pursue this problem as a [new GitHub issue](https://github.com/eclipse-theia/theia/issues/new?template=bug_report.md). Hopefully, it gets greater visibility.

**Reply by vvvjo** [reply]
@kittaakos Thank you for taking a deeper look. So I changed the line you mentioned and set it to `reveal: false` but I got the same result as before. Debug Console overrides the Output panel. My `internalConsoleOptions` are set to `openOnFirstSessionStart`. So, to clarify, when you set `reveal: false`, the Output panel showed rather than Debug Console after running `Hello World: Start Debugging` command?

I initially opened a Github issue, but it was closed and moved here. Would it be okay to open another one?

**Reply by kittaakos** [reply]
> but I got the same result as before

😕 Did you do it with `1.2.0`, `1.3.0`, or with the `HEAD`? Did you try with a Theia plug-in or a VS Code extension? They should behave the same, though.

> Would it be okay to open another one?

No,  let me take another look. I do not have much time for this issue lately, but it's a bug, so I'll try to get back to you with an update this week.

**Reply by kittaakos** [reply]
I have checked it again and compared the behavior with VS Code. The fundamental difference I see is that VS Code does not activate (and preserve the focus) of the _Debug Console_. Theia does, that's the problem and the not the `show` on the `OutputChannel`. Maybe we can align the Theia behavior with VS Code. Thoughts, @akosyakov? The easiest way to check how it works in Code is to make sure the _Debug Console_ is not active, and lunch a `index.js` app in debug mode. The app does not grab the focus on the _Debug Console_ although the output is dumped there. See the screencast.

![screencast 2020-07-01 11-09-27](https://user-images.githubusercontent.com/1405703/86226269-b503e100-bb8b-11ea-8e03-99672c673186.gif)

Long story short: it's not an Output issue and we need to change the debug behavior.

**Reply by akosyakov** [reply]
As far as I remember whether the debug console should be revealed is configured in launch configuration and it should be aligned already. Are you sure that you are using the very same launch configuration and debug preferences for testing?

**Reply by kittaakos** [reply]
> Are you sure that you are using the very same launch configuration and debug preferences for testing?

Yes. I used the same VS Code extension in both Theia and VS Code: https://github.com/kittaakos/output-channel-example

This is the way we start the debug session:
```ts
await vscode.debug.startDebugging(vscode.workspace.workspaceFolders![0], {
    "type": "node",
    "request": "launch",
    "name": "Launch Program",
    "program": "${workspaceFolder}/index.js",
});
```

**Reply by vvvjo** [reply]
> 😕 Did you do it with `1.2.0`, `1.3.0`, or with the `HEAD`? Did you try with a Theia plug-in or a VS Code extension? They should behave the same, though.

@kittaakos Apologies for missing to include this information. I recently tested again using `1.3.0` and `HEAD` setting `reveal: false` and it worked like you said! The Debug Console does not override the Output panel. In my last test, I was at HEAD but I think I missed transpiling before running `yarn start`. 

Thank you both for looking at this problem more in depth.

**Reply by vvvjo** [reply]
@akosyakov @kittaakos Following up to see if there are any updates.

**Reply by kittaakos** [reply]
> Maybe we can align the Theia behavior with VS Code. Thoughts, @akosyakov?
> Long story short: it's not an Output issue and we need to change the debug behavior.

I do not have any updates on this thread. I already added my findings.

**Reply by kittaakos** [reply]
VS Code [does not focus](https://github.com/microsoft/vscode/blob/e683ace9e5acadba0e8bde72d793cb2cb83e58a7/src/vs/workbench/contrib/debug/browser/debugService.ts#L487-L490) the _Debug Console_, Theia [does](https://github.com/eclipse-theia/theia/blob/68a3453ed1de6279dc8a44e7d846b92a0953ebe9/packages/debug/src/browser/debug-frontend-application-contribution.ts#L447-L452).

**Reply by akosyakov** [reply]
> Theia does.

It ~seems~ only reveals, not activate (focus).

**Reply by vvvjo** [reply]
@kittaakos @akosyakov So given that they both reveal, is the difference in behavior caused by how each handles asynchronous actions? Is there something that can be done to sync these behaviors? Thanks!

**Reply by vvvjo** [reply]
Open a PR for review, please take a look!
